src/Stratis.Bitcoin.Features.Consensus/CoinViews/DBreezeCoinView.cs

```diff
@@ -292,44 +296,70 @@ public RewindData GetRewindData(int height)
         }
 
         /// <inheritdoc />
-        public uint256 Rewind()
+        public uint256 Rewind(int? targetHeight = null)
         {
             uint256 res = null;
             using (DBreeze.Transactions.Transaction transaction = this.CreateTransaction())
             {
+                int currentHeight = this.GetRewindIndex(transaction);
+
+                targetHeight = targetHeight ?? (currentHeight - 1);
+
+                Guard.Assert(targetHeight >= 0 && targetHeight < currentHeight);
+
                 transaction.SynchronizeTables("BlockHash", "Coins", "Rewind");
-                if (this.GetRewindIndex(transaction) == 0)
+
+                if (targetHeight == 0)
                 {
                     transaction.RemoveAllKeys("Coins", true);
-                    this.SetBlockHash(transaction, this.network.GenesisHash);
+                    transaction.RemoveAllKeys("Rewind", true);
 
                     res = this.network.GenesisHash;
                 }
                 else
                 {
                     transaction.ValuesLazyLoadingIsOn = false;
 
-                    Row<int, byte[]> firstRow = transaction.SelectBackward<int, byte[]>("Rewind").FirstOrDefault();
-                    transaction.RemoveKey("Rewind", firstRow.Key);
-                    var rewindData = this.dBreezeSerializer.Deserialize<RewindData>(firstRow.Value);
-                    this.SetBlockHash(transaction, rewindData.PreviousBlockHash);
+                    var keysToRemove = new List<int>();
+                    var changes = new Dictionary<uint256, Coins>();
```

Comment:
I'm wondering whether or not we shouldn't have 2 dictionaries here. Just so that we are more explicit as to what we are going to be changing? I don't mind either way but it might read a bit better and be more self-explanatory.

Reply:
The single dictionary more accurately simulates the legacy insert/delete into the single table.